### ISSUE TITLE
Support drag and drop on wayland

### ISSUE THREAD

**Issue by thiolliere** [rust-windowing/winit#1881]
Drag and drop windows events are currently not supported on wayland.

I wonder how it is possible to implement it considering smithay-client-toolkit contains some helper for it: DataDevice.

**Comment by tfachmann** [comment]
Can we use https://github.com/Smithay/wayland-rs i.e. wayland-client for implementation? Smithay's wayland client toolkit builds on top of it, so it should be a seamless integration.

The `wl_data_device_manager` might be what we need to implement drag and drop in wayland: `https://smithay.github.io/wayland-rs/wayland_client/protocol/wl_data_device_manager/index.html

**Comment by WhyNotHugo** [comment]
The [`DataDevice`](https://docs.rs/smithay-client-toolkit/latest/smithay_client_toolkit/data_device/struct.DataDevice.html) helper can be used by itself too. smithay toolkit isn't an all-or-nothing kind of thing; it's perfectly feasible to cherry pick individual helpers.

**Comment by kchibisov** [comment]
@WhyNotHugo gnome breaks if you bind multiple data devices(it'll use only one of them), so unless winit has clipboard adding drag and drop is not possible.

**Comment by WhyNotHugo** [comment]
Upon further inspection (I was trying to use DataDevice directly in my application) it seems that it's been dropped in sctk-master anyway, so it's best not to rely on it.

I'm not sure what the correct behaviour is when binding to a global twice, but I'm pretty sure it's not something that should be done anyway.

**Comment by kchibisov** [comment]
>I'm not sure what the correct behaviour is when binding to a global twice, but I'm pretty sure it's not something that should be done anyway.

If something is missing is stck master, it's only because it wasn't ported yet. SCTK had abstractions for dnd and clipboard(it's the same thing) for a long time. We have a PR adding DND already for Wayland, it just breaks clipboard on gnome, because gnome doesn't allow client to bind multiple times to data device.

**Comment by kchibisov** [comment]
>Good question. Maybe the Wayland protocol spec could use some clarification on this? Normally binding something multiple times is perfectly fine and works correctly (and may be needed when multiple libraries are making Wayland calls), so without text in the spec about this the Gnome behavior sounds incorrect.

That works fine on sway for example, it just gnome is special and doesn't bother to fix anything. I bet it still sends `modifiers` events incorrectly and I have to carry workarounds for gnome to ignore its non-compliance to wayland protocol.

**Comment by WhyNotHugo** [comment]
Ah, thanks for pointing out the new DataDevice implementation in sctk, I'd missed it. This seems to bind the data device global only once for both clipboard and drag-and-drop, so should be usable here, right?

Looks like the GNOME bug was fixed: https://gitlab.gnome.org/GNOME/mutter/-/merge_requests/1253

**Comment by ids1024** [comment]
> This seems to bind the data device global only once for both clipboard and drag-and-drop, so should be usable here, right?

I believe the issue is that winit doesn't currently provide clipboard support (https://github.com/rust-windowing/winit/pull/2156). It looks like toolkits and such in Rust are currently mostly using `smithay-clipboard` (directly or via a crate that abstracts multiple platforms). But another library like that will bind its own copy of the data device.

So I guess to fix it on the version of Mutter with this issue, it would be necessary to also add clipboard support to winit, and then software updating to the new version of winit would need to update their clipboard support to use winit's implementation. Which for the most part is probably how it should be done anyway.

Though maybe not an issue if it was fixed in Gnome a couple years ago? Not sure what release it made it into of if it was backported. Or if this is any issue in any other compositor.

**Comment by WhyNotHugo** [comment]
GitLab indicates that the fix made it to the `gnome-3-36-backport` branch: https://gitlab.gnome.org/rmader/mutter/-/commit/1ec91cc8eef1db353d852fc4c212cae3a9cb1aee

**Comment by ids1024** [comment]
Okay, so that should be on Ubuntu 20.04 (the previous LTS) and the like.

That sounds like it should generally cover current Gnome users that are using Wayland.

**Comment by cdata** [comment]
#3367 is completed as of almost a year ago. Would it be possible to note the latest blockers to fixing this issue?

**Comment by zinifier** [comment]
I tried applying #4009 and it seems to fix the problem in my example application.

**Comment by timon-schelling** [comment]
Anything new on this?
We need this in the upcoming Graphite Desktop App.
Are there any blockers or anything else we might be able to help with?

**Comment by kchibisov** [comment]
The question is whether you'd need it in winit 0.30 or 0.31. Assuming that we're talking about dnd in the form it is right now which is very limited.

**Comment by timon-schelling** [comment]
Depends on the timeline for 0.31. Will the dnd system be reworked in 0.31? If yes, does that mean Wayland support will not come in 0.30 because you prioritize the long term solution?

We are trying to release soon, but Linux/Wayland features missing will not be blocking.

**Comment by lanastara** [comment]
Are there any plans to include this in 0.31 now that the dnd system has been reworked?

**Comment by Third-Thing** [comment]
1. On 2022-8-12 @SludgePhD opened [DnD PR 2429 on winit](https://github.com/rust-windowing/winit/pull/2429); maintainer @kchibisov engaged but the conversation went silent on both ends
2. On 2024-1-11 in this thread kchibisov pointed to something related to #3367 as blocking DnD
3. On 2024-1-26 that issue was marked complete and split into 5 issues (see the last comment); four are still open but it's not clear how any of them are blocking wayland DnD
4. On 2024-3-25 @wash2 opened [DnD PR 57 on smithay-clipboard](https://github.com/Smithay/smithay-clipboard/pull/57); lead maintainer kchibisov made no comment, but later informed me that smithay-clipboard is the wrong approach and they won't be merging new features.
5. On 2024-11-26 @PeakKS opened [DnD PR 4009 on winit](https://github.com/rust-windowing/winit/pull/4009); kchibisov linked to 2429 and said "it all should still apply."; what that meant is unclear
6. The same day, PeakKS pushed changes and said "I think the functionality is correct now"; no maintainers ever responded
7. On 2025-1-28 A DnD API rework got merged #4079
8. On 2025-8-12 kchibisov asserted on 2429 that "Either I will be the one updating this version... or you can pick it up..."; the original author indicated a lack of interest
9. On 2026-3-6 @joeleaver opened [DnD PR 4504 on winit](https://github.com/rust-windowing/winit/pull/4504)
10. On 2026-4-4 kchibisov closed PR 4504 without comment. No maintainer left a single comment on the PR at any point.

Four independent contributors have implemented Wayland DnD since this issue was opened 5 years ago. And the community still doesn't have it.

- Why was PeakKS's pull request ignored?
- Why did kchibisov favor the older pull request as something they would work on?
- Why was that work not done in the last 8 months?
- Why was the new pull request ignored for a month and then closed without comment?

There's a lot of people that were hoping this would be in v0.31.0. What _exactly_ is blocking this? And how can we help?

**Comment by uonr** [comment]
Is https://github.com/rust-windowing/winit/pull/4571 resolves this issue?

**Comment by ogoffart** [comment]
yes
